tests/test_server_pool.py

```diff
@@ -340,33 +494,87 @@ def simulate_all_and_collect_stats(self):
 
         js_data = []
         for testname, spec in specs.items():
-            print(f'Running {testname}...', end='', flush=True)
+            print(f'Running {testname}...')
             js_data.append(
                 asyncio.run(
                     self.simulate_and_collect_stats(testname, spec)))
-            print('OK')
 
         html = string.Template(HTML_TPL).safe_substitute(
             DATA=json.dumps(js_data))
+        score = int(round(statistics.fmean(
+            sim['runs'][0]['score'] for sim in js_data
+        )))
+
+        if os.environ.get("SIMULATION_CI"):
+            sha = os.environ.get('GITHUB_SHA')
+            path = f'reports/{sha}.html'
+            report_path = f'pool-simulation/{path}'
+            i = 1
+            while os.path.exists(report_path):
+                path = f'reports/{sha}-{i}.html'
+                report_path = f'pool-simulation/{path}'
+                i += 1
+            with open(report_path, 'wt') as f:
+                f.write(html)
+            try:
+                with open(f'pool-simulation/reports.json') as f:
+                    reports = json.load(f)
+            except Exception:
+                reports = []
+            reports.insert(0, {
+                'path': path,
+                'sha': sha,
+                'ref': os.environ.get('GITHUB_REF'),
+                'num_simulations': len(js_data),
+                'qos_score': score,
+                'datetime': str(datetime.datetime.now()),
+            })
+            with open(f'pool-simulation/reports.json', 'wt') as f:
+                json.dump(reports[:CI_MAX_REPORTS], f)
+            with open(f'pool-simulation/reports-archive.json', 'at') as f:
+                for report in reports[:CI_MAX_REPORTS - 1:-1]:
+                    print('Removing outdated report:', report['path'])
+                    json.dump(report, f)
+                    print(file=f)
+                    try:
+                        os.unlink(report['path'])
+                    except OSError as e:
+                        print('ERROR:', e)
+
+        else:
+            if not os.path.exists('tmp'):
+                os.mkdir('tmp')
+            with open(f'tmp/connpool.html', 'wt') as f:
+                f.write(html)
+            now = int(datetime.datetime.now().timestamp())
+            with open(f'tmp/connpool_{now}.html', 'wt') as f:
+                f.write(html)
 
-        if not os.path.exists('tmp'):
-            os.mkdir('tmp')
-        with open(f'tmp/connpool.html', 'wt') as f:
-            f.write(html)
-        now = int(datetime.datetime.now().timestamp())
-        with open(f'tmp/connpool_{now}.html', 'wt') as f:
-            f.write(html)
+        print('Final QoS score:', score)
+        return score
 
 
 class TestServerConnpoolSimulation(SimulatedCase):
 
     def test_server_connpool_1(self):
+        def calc_score(sim: Simulation):
+            return self.calc_group_score_with_weight(
```

Comment:
Yeah, just glancing at it I've no idea how to use it. Docstrings would be hugely appreciated!

Reply:
Will do!